hessian2/basic_codec/string_codec.cc

```diff
@@ -4,6 +4,42 @@ namespace Hessian2 {
 
 namespace {
 constexpr size_t STRING_CHUNK_SIZE = 32768;
+
+static const size_t UTF_8_CHAR_LENGTHS[] = {1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+                                            1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0,
+                                            0, 0, 2, 2, 2, 2, 3, 3, 4, 0};
+
+template <bool WITH_PER_CHUNK>
+int64_t getUtf8StringLength(const absl::string_view &out,
+                            std::vector<uint64_t> &per_chunk_raw_size) {
+  size_t utf_len = 0;
+  size_t i = 0;
+  size_t next_chunk_size = 0;
+
+  for (; i < out.size();) {
+    auto code = static_cast<uint8_t>(out[i]);
+    auto size = UTF_8_CHAR_LENGTHS[code >> 3];
+
+    if (size == 0) {
+      return -1;
+    }
+
+    utf_len++;
+    i += size;
+
+    if constexpr (WITH_PER_CHUNK) {
```

Comment:
It's just a matter of avoiding branching prediction, using `constexpr` to get the compiler to optimize it, right? Is it possible to use likely?

Reply:
`if constexpr` allows the compiler to determine the result of the judgment and the logic that needs to be executed during compilation. There will be no judgment at runtime. If you can completely determine the branch that needs to be executed during compilation, then `if constexpr` has a better optimization effect than `likely`.